teps/0117-tekton-results-logs.md

```diff
@@ -0,0 +1,350 @@
+
+Example configuration file:
+
+```yaml
+database:
+  address: <url>
+  name: <db-name>
+  sslMode: verify-full
+  ...
+logStorage:
+  enabled: true
+  backendType: File | S3 | ... # Declare which back end type should be used
+  file:
+    rootDir: /var/tekton-results/logs # Directory where logs should be persisted
+  s3: {} # S3 configuration
+```
+
+The APIServer codebase includes "drivers" that let it store and retrieve log
+data for the respective backend. These will be provided in-tree with the
+Results code.
+
+## Design Evaluation
+
+- `TaskRunLog` data uses Kubernetes API conventions (even though this is not a
+  k8s object)
+
+### Reusability
+
+- This proprosal provides a common solution to a problem that is presently
+  addressed with ad-hoc, one-off solutions.
+- This proposal reuses the Records data table to store log persistence
+  information. It does not propose the creation of a new database table.
+
+### Simplicity
+
+- This proposal tries to keep things simple by centralizing most configuration
+  in the Results APIServer.
+- This proposal provides a single API endpoint for retrieving TaskRun logs from
+  persistent storage.
+- This proposal provides "in-tree" mechanisms for extending supported log
+  storage backends.
+- Because `TaskRunLog` data exists as a results Record, the following implicit
+  behavior may need to be addressed:
+  - Uploading logs requires a `TaskRunLog` record to be created in the first
+    place.
+  - Deleting a record of type `TaskRunLog` - should this cascade deletion to
+    the log itself if the data is persisted by Results?
+
+### Flexibility
+
+- To forward logs, this couples Results to the tkn CLI. This could be mitigated
+  by refactoring the `tkn logs` libraries to Pipelines or another common
+  repository.
+- As more backends are introduced, they introduce additional dependencies for
+  their respective SDKs. This is especially true for cloud provider object
+  storage.
+
+### Conformance
+
+- End users do not need to understand the Tekton API implementation details.
+- No core Tekton APIs are changed.
+
+### User Experience
+
+- There is a tkn plugin for Results `tkn-results`, which can be enhanced to
+  fetch logs from Results.
+- Dashboard can also be enhanced to retrieve logs from Results as a fallback.
+  The Dashboard already has a
+  [fallback mechanism for s3](https://github.com/tektoncd/dashboard/blob/main/docs/walkthrough/walkthrough-logs.md#installing-minio-as-an-object-storage-solution)
+
+### Performance
+
+- This may impact the rate at which Results can reconcile completed TaskRuns,
+  especially if the task steps produce copious logs.
+
+### Risks and Mitigations
+
+Primary risk is unauthorized access to logs from a TaskRun. This is mitigated
+through subject access reviews and Kubernetes RBAC, which is built into the
+Results APIServer.
+
+There is also a denial of service risk if a TaskRun emits significant volumes
```

Comment:
have you considered integrating with things like [fluentbit](https://fluentbit.io/) that can already forward logs instead of writing our own streaming code?

Reply:
Generally, clusters have something like fluentbit/fluentd enabled. But in certain situations, setting up of fluent-bit will entail further configurations because fluentbit/d parses every pod in clusters and is available in every node. So this minimal approach is good.

Reply:
That is the intent of the `External` category in `status.storage`. This tells the watcher that an external process (fluend, fluentbit, etc.) is forwarding the logs. The APIServer would then need an appropriate backend to retrieve the logs.